extractor/filesystem/os/opkg/opkg.go

```diff
@@ -0,0 +1,285 @@
+
+// Package opkg extracts packages from opkg database.
+package opkg
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/textproto"
+	"path/filepath"
+	"strings"
+
+	"github.com/google/osv-scalibr/extractor"
+	"github.com/google/osv-scalibr/extractor/filesystem"
+	"github.com/google/osv-scalibr/extractor/filesystem/internal/units"
+	"github.com/google/osv-scalibr/extractor/filesystem/os/osrelease"
+	"github.com/google/osv-scalibr/log"
+	"github.com/google/osv-scalibr/plugin"
+	"github.com/google/osv-scalibr/purl"
+	"github.com/google/osv-scalibr/stats"
+	"golang.org/x/text/cases"
+	"golang.org/x/text/language"
+)
+
+const (
+	// Name is the unique name of this extractor.
+	Name = "os/opkg"
+
+	// defaultMaxFileSizeBytes is the maximum file size an extractor will unmarshal.
+	// If Extract gets a bigger file, it will return an error.
+	defaultMaxFileSizeBytes = 10 * units.MiB
+
+	// defaultIncludeNotInstalled is the default value for the IncludeNotInstalled option.
+	defaultIncludeNotInstalled = false
+)
+
+// Config is the configuration for the Extractor.
+type Config struct {
+	// Stats is a stats collector for reporting metrics.
+	Stats stats.Collector
+	// MaxFileSizeBytes is the maximum file size this extractor will unmarshal. If
+	// `FileRequired` gets a bigger file, it will return false,
+	MaxFileSizeBytes int64
+	// IncludeNotInstalled includes packages that are not installed
+	// (e.g. `deinstall`, `purge`, and those missing a status field).
+	IncludeNotInstalled bool
+}
+
+// DefaultConfig returns the default configuration for the OPKG extractor.
+func DefaultConfig() Config {
+	return Config{
+		MaxFileSizeBytes:    defaultMaxFileSizeBytes,
+		IncludeNotInstalled: defaultIncludeNotInstalled,
+	}
+}
+
+// Extractor extracts packages from OPKG files.
+type Extractor struct {
+	stats               stats.Collector
+	maxFileSizeBytes    int64
+	includeNotInstalled bool
+}
+
+// New returns a OPKG extractor.
+//
+// For most use cases, initialize with:
+// ```
+// e := New(DefaultConfig())
+// ```
+func New(cfg Config) *Extractor {
+	return &Extractor{
+		stats:               cfg.Stats,
+		maxFileSizeBytes:    cfg.MaxFileSizeBytes,
+		includeNotInstalled: cfg.IncludeNotInstalled,
+	}
+}
+
+// Config returns the configuration of the extractor.
+func (e Extractor) Config() Config {
+	return Config{
+		Stats:               e.stats,
+		MaxFileSizeBytes:    e.maxFileSizeBytes,
+		IncludeNotInstalled: e.includeNotInstalled,
+	}
+}
+
+// Name of the extractor.
+func (e Extractor) Name() string { return Name }
+
+// Version of the extractor.
+func (e Extractor) Version() int { return 0 }
+
+// Requirements of the extractor.
+func (e Extractor) Requirements() *plugin.Capabilities { return &plugin.Capabilities{} }
+
+// FileRequired checks if the specified file matches OPKG status file pattern.
+func (e Extractor) FileRequired(api filesystem.FileAPI) bool {
+	path := api.Path()
+	if !fileRequired(path) {
```

Comment:
We should inline the path matching instead of having it in a separate function. Otherwise we'd do an additional function call for every single file on the filesystem which can increase the runtime significantly